stdlib/public/core/Flatten.swift.gyb

```diff
@@ -313,40 +258,51 @@ public struct ${Collection}<
   /// reachable from `startIndex` by zero or more applications of
   /// `successor()`.
   public var endIndex: Index {
-    return ${Index}(_endIndexOfFlattened: _base)
+    return ${Index}(_base.endIndex, nil)
   }
 
   // TODO: swift-3-indexing-model - add docs
   @warn_unused_result
   public func next(i: Index) -> Index {
-    fatalError("FIXME: swift-3-indexing-model implement")
+    // FIXME: swift-3-indexing-model: range checks?
+    let nextInner = _base[i._outer].next(i._inner!)
+    if _fastPath(nextInner != _base[i._outer].endIndex) {
+      return ${Index}(i._outer, nextInner)
+    }
+
+    var nextOuter = _base.next(i._outer)
+    while nextOuter != _base.endIndex {
+      let innerCollection = _base[nextOuter]
+      if !innerCollection.isEmpty {
+        return ${Index}(nextOuter, innerCollection.startIndex)
+      }
+      _base._nextInPlace(&nextOuter)
+    }
+
+    return endIndex
   }
 
 % if traversal == 'Bidirectional':
   // TODO: swift-3-indexing-model - add docs
   @warn_unused_result
   public func previous(i: Index) -> Index {
-    fatalError("FIXME: swift-3-indexing-model implement")
-  }
-% end
-  
-  // TODO: swift-3-indexing-model - add docs
-  @warn_unused_result
-  public func advance(i: Index, by n: IndexDistance) -> Index {
-    fatalError("FIXME: swift-3-indexing-model implement")
-  }
-  
-  // TODO: swift-3-indexing-model - add docs
-  @warn_unused_result
-  public func advance(i: Index, by n: IndexDistance, limit: Index) -> Index {
-    fatalError("FIXME: swift-3-indexing-model implement")
-  }
+    // FIXME: swift-3-indexing-model: range checks?
```

Comment:
I think we don't need any range checks.  Base collections should be doing their own.